molgenis-app-gavin2/src/main/java/org/molgenis/app/gavin/GavinController.java

```diff
@@ -0,0 +1,191 @@
+  @GetMapping(value = "/run/{id}")
+  public GavinRunResponse get(@PathVariable String id) {
+    return GavinRunResponse.create(gavinService.get(id));
+  }
+
+  @PostMapping(value = "/run/{id}/start")
+  public ResponseEntity start(@PathVariable String id) {
+    gavinService.start(id);
+    return new ResponseEntity(HttpStatus.OK);
+  }
+
+  @PostMapping(value = "/run/{id}/finish")
+  public ResponseEntity finish(
+      @PathVariable String id,
+      @RequestParam MultipartFile outputFile,
+      @RequestParam String log,
+      HttpServletRequest httpServletRequest)
+      throws IOException, ServletException {
+    gavinService.finish(id, log, httpServletRequest);
+    return new ResponseEntity(HttpStatus.OK);
+  }
+
+  @PostMapping(value = "/run/{id}/fail")
+  public ResponseEntity start(@PathVariable String id, @RequestParam String log) {
+    gavinService.fail(id, log);
+    return new ResponseEntity(HttpStatus.OK);
+  }
+
+  @RunAsSystem
+  @GetMapping(value = "/run/{id}/download/output", produces = APPLICATION_OCTET_STREAM_VALUE)
+  public FileSystemResource downloadOutputFile(
+      HttpServletResponse response, @PathVariable(value = "id") String id) {
+    FileMeta outputFile = gavinService.get(id).getOutputFile();
+    return prepareDownload(response, outputFile);
+  }
+
+  @RunAsSystem
+  @GetMapping(value = "/run/{id}/download/input", produces = APPLICATION_OCTET_STREAM_VALUE)
+  public FileSystemResource downloadInputFile(
+      HttpServletResponse response, @PathVariable(value = "id") String id) {
+    FileMeta errorFile = gavinService.get(id).getFilteredInputFile();
+    return prepareDownload(response, errorFile);
+  }
+
+  @RunAsSystem
+  @GetMapping(value = "/run/{id}/download/error", produces = APPLICATION_OCTET_STREAM_VALUE)
+  public FileSystemResource downloadErrorFile(
+      HttpServletResponse response, @PathVariable(value = "id") String id) {
+    FileMeta errorFile = gavinService.get(id).getDiscardedInputFile();
+    return prepareDownload(response, errorFile);
+  }
+
+  private FileSystemResource prepareDownload(HttpServletResponse response, FileMeta fileMeta) {
+    if (fileMeta == null) {
+      response.setStatus(HttpServletResponse.SC_NO_CONTENT);
+      return null;
+    }
+
+    File file = fileStore.getFile(fileMeta.getId());
+    response.setHeader(
+        "Content-Disposition", format("attachment; filename=\"%s\"", fileMeta.getFilename()));
+    return new FileSystemResource(file);
+  }
+
+  /**
+   * Removes files from gavin runs that have finished more than 24 hours ago. Does not remove the
+   * GavinRun entities themselves so that they can be used for usage statistics.
+   */
+  @RunAsSystem
+  @Scheduled(fixedRate = 5 * 60 * 1000)
+  void cleanUp() {
+    LOG.debug("Starting cleanup routine of expired GavinRuns");
+
+    Stream<GavinRun> completedGavinRuns =
+        dataService
+            .query(GAVIN_RUN, GavinRun.class)
+            .in(STATUS, asList(Status.FAILED, Status.SUCCESS))
```

Comment:
How about Status.CANCELED as used in the job framework?

Reply:
The scheduled job could mark jobs as CANCELED or FAILED when they haven't reached another end state in a certain timespan (e.g. 24 hours) and clean up resources to prevent resources staying available indefinitely.